### ISSUE TITLE
Field injection for binding is a mistake

### ISSUE THREAD

**Issue by chkal** [eclipse-ee4j/mvc-api#37]
<a href="https://github.com/graemerocher"><img src="https://avatars2.githubusercontent.com/u/66626?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [graemerocher](https://github.com/graemerocher)**
_Wednesday Nov 14, 2018 at 14:01 GMT_
_Originally opened as https://github.com/mvc-spec/mvc-spec/issues/206_

----

Reading the spec, it seems that the pattern that is encouraged is this:

```java
@Inject
    private User user;

    @GET
    @Controller
    public String hello(@QueryParam("name") String name) {
        user.setName(name);
        return "hello.jsp";
    }
```

IMO this is a mistake looking to the future. There are several problems with this:

1. Thread safety if the controller is singleton. The controller would not automatically be thread safe, which would force the introduction of proxies that use request scope/ thread local, which impacts performance and memory consumption
2. Requirement for reflection - frameworks would be required to use reflection to inject `User user` using a field based approach only

Fields should be used for dependency injected components and not the model. It would be better written as:

```java
    @GET
    @Controller
    public String hello(@Model User user, @QueryParam("name") String name) {
        user.setName(name);
        return "hello.jsp";
    }
```

This has the advantages:

1. Reflection would not necessary be required (for example in Micronaut)
2. There are no thread safety concerns with singleton/application scoped controllers.

This also aligns closer with how the most popular MVC (Spring MVC) works.

Relates to #16

**Comment by chkal** [comment]
<a href="https://github.com/chkal"><img src="https://avatars2.githubusercontent.com/u/159245?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [chkal](https://github.com/chkal)**
_Sunday Feb 03, 2019 at 13:38 GMT_

----

Hi @graemerocher,

thanks a lot for reviewing the spec and for providing feedback. That's always valuable and helps us a lot. And sorry for the delayed response. We are very busy building the TCK and the RI which is a lot of work.

First of all I would like to clarify one thing. MVC 1.0 targets Java/Jakarta EE and therefore integrates very closely with other parts of the platform. As CDI is becoming the core component model of the platform, we decided to build MVC on top of it. The great thing about this is that developers use a consistent component model across the platform and that many problems are already solved for us, like:

- Dependency injection in different flavors (constructor, setter and field injection)
- Creation for proxies to handle method interception and scope differences

See my thoughts about your concerns below:

> 1. Thread safety if the controller is singleton. The controller would not automatically be thread safe, which would force the introduction of proxies that use request scope/ thread local, which impacts performance and memory consumption

I don't see any good reason why a controller should be a singleton. Using the request scope for MVC controllers and JAX-RS resources totally makes sense to me. They process individual request, so why not create one instance per request? Object creation is extremely cheap compared to database queries which are typically performed within the same request.

And as mentioned above, proxy creation is out of scope for us as this is part of the core component model provided by CDI. Yes, there will be some overhead for proxies but today's CDI implementations are highly optimized to reduce this overhead, so I don't think that this is really an issue.

> 2. Requirement for reflection - frameworks would be required to use reflection to inject User user using a field based approach only

CDI supports various ways to perform dependency injection. So developers can use constructor injection, property/setter injection and field injection. So it is not "only" field based as you describe it. And yes, reflection is required, but that is done by CDI and already solved for us.

However, I would really be happy to also get injection for controller methods as described in #16. But that's something JAX-RS has to provide. I'm very confident that we can provide this with MVC 1.1 as JAX-RS is targeting improved CDI integration for the next release.

To sum up: It is very important to understand that MVC 1.0 targets Java/Jakrata EE and therefore integrates with core technologies of the platform. IMO this is a good thing, although the resulting framework will of course be much different from standalone approaches like Micronaut.

**Comment by erdlet** [comment]
@chkal seems like a zombie issue to me since this is no MVC topic. Shall we close it?

**Comment by chkal** [comment]
I agree. Let's close this one. I don't think that there is much we can do about it, as we rely on CDI.
